### DISCUSSION TITLE
Regarding use of RGCN

### DISCUSSION THREAD

**Discussion by urmi22** [pyg-team/pytorch_geometric#3983] [Q&A]
I have attached pictorial representaion of a model here. In that model I want to use RGCN-CONV, where input is a graph (graph consist of two types of nodes : documents and concepts; and three types of relational edges : document to document, document to concept and concept to concept) and output is a prerequisite prection between two concepts. I just need to know how I can implement this entire model using RGCN.
each node type and edges have features.
![Selection_001](https://user-images.githubusercontent.com/34470785/151974436-4668e35f-07cd-4b2b-b160-a21065796d03.png)
paper link. https://aclanthology.org/2021.naacl-main.164.pdf

**Comment by rusty1s** [comment]
I suggest to read up how PyG handles heterogeneous data [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html). Let me know if you have any questions.

**Reply by urmi22** [reply]
Thanks @rusty1s for your reply!

I'll try to explain the problem I facing for now. My heterogeneous graph metedata looks as follows- 

```
HeteroData(
  concept={ x=[411, 50] },
  document={ x=[599, 50] },
  (document, contains, concept)={
    edge_index=[2, 1237],
    edge_attr=[1237, 1]
  },
  (document, related-to, document)={
    edge_index=[2, 358801],
    edge_attr=[358801, 1]
  },
  (concept, similar-to, concept)={
    edge_index=[2, 3118],
    edge_attr=[3118, 1],
    y=[3118]
  }
)
```

This I have created following the tutorial you have pointed out. I have defined the RGCN model as follows to extract `node representation`. The code looks like this - 

```
 class Model(nn.Module):
	def __init__(self):
		super(Model,self).__init__()
		self.rgcn_layer = RGCNConv(in_channels=50, out_channels=120, num_relations=3, num_bases=3)
		
	def forward(self, x, edge_index, edge_types):
		x = self.rgcn_layer(x, edge_index, edge_types)
		return x
model = to_hetero(Model(), data.metadata(), aggr='sum')
```

However, if I call the `forward` method, I get the following error: 

```
>> output = model(data.x_dict, data.edge_index_dict, data.edge_types)
*** AttributeError: 'List' object has no attribute 'get'
```


Can you please point me out the mistake I am making?

**Reply by rusty1s** [reply]
I see. I don't think it makes sense to combine both `to_hetero` and `RGCNConv`. You either need to convert your heterogeneous graph into a "typed" homogeneous one via `data.to_homgeneous()` from where you can then make use of `RGCNConv`. Alternatively, you define a regular homegeneous GNN (e.g., via `SAGEConv`), and then apply `to_hetero` on top of this model.

**Reply by urmi22** [reply]
@rusty1s Thanks! I also had this similar doubt. I'll try what you have suggested and let you know if I face any other issues.
